server/src/websocket/server.rs

```diff
@@ -389,352 +389,526 @@ impl ChatServer {
+            .await
+        }
+        UserOperation::GetPosts => {
+          do_user_operation::<GetPosts>(pool, rate_limiter, addr, msg.id, ip, user_operation, data)
+            .await
+        }
+        UserOperation::EditPost => {
+          do_user_operation::<EditPost>(pool, rate_limiter, addr, msg.id, ip, user_operation, data)
+            .await
+        }
+        UserOperation::CreatePostLike => {
+          do_user_operation::<CreatePostLike>(
+            pool,
+            rate_limiter,
+            addr,
+            msg.id,
+            ip,
+            user_operation,
+            data,
+          )
+          .await
+        }
+        UserOperation::SavePost => {
+          do_user_operation::<SavePost>(pool, rate_limiter, addr, msg.id, ip, user_operation, data)
+            .await
+        }
+
+        // Comment ops
+        UserOperation::CreateComment => {
+          do_user_operation::<CreateComment>(
+            pool,
+            rate_limiter,
+            addr,
+            msg.id,
+            ip,
+            user_operation,
+            data,
+          )
+          .await
+        }
+        UserOperation::EditComment => {
+          do_user_operation::<EditComment>(
+            pool,
+            rate_limiter,
+            addr,
+            msg.id,
+            ip,
+            user_operation,
+            data,
+          )
+          .await
+        }
+        UserOperation::SaveComment => {
+          do_user_operation::<SaveComment>(
+            pool,
+            rate_limiter,
+            addr,
+            msg.id,
+            ip,
+            user_operation,
+            data,
+          )
+          .await
+        }
+        UserOperation::GetComments => {
+          do_user_operation::<GetComments>(
+            pool,
+            rate_limiter,
+            addr,
+            msg.id,
+            ip,
+            user_operation,
+            data,
+          )
+          .await
+        }
+        UserOperation::CreateCommentLike => {
+          do_user_operation::<CreateCommentLike>(
+            pool,
+            rate_limiter,
+            addr,
+            msg.id,
+            ip,
+            user_operation,
+            data,
+          )
+          .await
+        }
+      }
     }
   }
 }
 
+async fn do_user_operation<'a, Data>(
+  pool: Pool<ConnectionManager<PgConnection>>,
+  rate_limiter: RateLimit,
+  chatserver: Addr<ChatServer>,
+  id: ConnectionId,
+  ip: IPAddr,
+  op: UserOperation,
+  data: &str,
+) -> Result<String, Error>
+where
+  for<'de> Data: Deserialize<'de> + 'a,
+  Oper<Data>: Perform,
+{
+  let ws_info = WebsocketInfo {
+    chatserver,
+    id: Some(id),
+  };
+
+  let data = data.to_string();
+  let op2 = op.clone();
+  let fut = async move {
+    let parsed_data: Data = serde_json::from_str(&data)?;
+    let res = Oper::new(parsed_data).perform(pool, Some(ws_info))?;
+    to_json_string(&op, &res)
+  };
```

Comment:
Is this just in anticipation of making perform async? I'm not exactly sure what the async move here is for. 

edit: oh nm, its so that you can wrap the future, got it.